rpcs3/rpcs3qt/update_manager.cpp

```diff
@@ -0,0 +1,541 @@
+	if (!func(*this, data, automatic))
+	{
+		if (!automatic)
+			QMessageBox::warning(m_progress_dialog, tr("Auto-updater"), tr("An error occured during the auto-updating process.\nCheck the log for more information."));
+
+		m_progress_dialog->close();
+	}
+
+	return true;
+}
+
+bool update_manager::handle_json(const QByteArray& data, bool automatic)
+{
+	const QJsonObject json_data = QJsonDocument::fromJson(data).object();
+	int return_code             = json_data["return_code"].toInt(-255);
+
+	bool hash_found = true;
+
+	if (return_code < 0)
+	{
+		std::string error_message;
+		switch (return_code)
+		{
+		case -1: error_message = "Hash not found(Custom/PR build)"; break;
+		case -2: error_message = "Server Error - Maintenance Mode"; break;
+		case -3: error_message = "Server Error - Illegal Search"; break;
+		case -255: error_message = "Server Error - Return code not found"; break;
+		default: error_message = "Server Error - Unknown Error"; break;
+		}
+
+		if (return_code != -1)
+			LOG_ERROR(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+		else
+			LOG_WARNING(GENERAL, "[Auto-updater] error: %s return code: %d", error_message, return_code);
+
+		// If a user clicks "Check for Updates" with a custom build ask him if he's sure he wants to update to latest version
+		if (!automatic && return_code == -1)
+		{
+			if (QMessageBox::question(m_progress_dialog, tr("Auto-updater"), tr("You're currently using a custom or PR build.\n\nDo you want to update to the latest official rpcs3 version?"),
+			        QMessageBox::Yes | QMessageBox::No) == QMessageBox::No)
+			{
+				m_progress_dialog->close();
+				return true;
+			}
+
+			hash_found = false;
+		}
+		else
+		{
+			return false;
+		}
+	}
+
+	const auto& latest = json_data["latest_build"];
+	if (!latest.isObject())
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Json doesn't contain latest_build section");
+		return false;
+	}
+
+	QString os;
+
+#ifdef _WIN32
+	os = "windows";
+#elif defined(__linux__)
+	os = "linux";
+#else
+	LOG_ERROR(GENERAL, "[Auto-updater] Your os isn't currently supported by the auto-updater");
+	return false;
+#endif
+
+	// Check that every bit of info we need is there
+	if (!latest[os].isObject() || !latest[os]["download"].isString() || !latest[os]["size"].isDouble() || !latest[os]["checksum"].isString() || !latest["version"].isString() ||
+	    !latest["datetime"].isString() ||
+	    (hash_found && (!json_data["current_build"].isObject() || !json_data["current_build"]["version"].isString() || !json_data["current_build"]["datetime"].isString())))
+	{
+		LOG_ERROR(GENERAL, "[Auto-updater] Some information seems unavailable");
+		return false;
+	}
+
+	if (hash_found && return_code == 0)
+	{
+		LOG_SUCCESS(GENERAL, "[Auto-updater] RPCS3 is up to date!");
+		if (!automatic)
+			QMessageBox::information(m_progress_dialog, tr("Auto-updater"), tr("Your version is already up to date!"));
+
+		m_progress_dialog->close();
```

Comment:
You should close this before spawning the MessageBox. Having the progress dialog lingering in the background and flashing up before continuation is not necessary at this point.